File: 01-Patterns/JavaScript/02-patterns.js
// *****
// *   *
// *   *
// *   *
// *****
const hollowBox = () => {
  for (let i = 0; i < 5; i++) {
    for (let j = 0; j < 5; j++) {
      if (i === 0 || i === 4 || j === 0 || j === 4) {
        process.stdout.write("*");
      } else {
        process.stdout.write(" ");
      }
    }
    console.log();
  }
};

// hollowBox();

// *   *
//  * *
//   *
//  * *
// *   *
const cross = () => {
  let dif = 4;
  for (let i = 1; i < 6; i++) {
    for (let j = 1; j < 6; j++) {
      if (i === j || j - i === dif) {
        process.stdout.write("*");
      } else {
        process.stdout.write(" ");
      }
    }
    console.log();
    dif = dif - 2;
  }
};

// cross();

//     *
//    **
//   ***
//  ****
const invertedAndRotatedPyramid = () => {
  space = 4;
  for (let i = 0; i < 5; i++) {
    for (let j = 0; j < 5; j++) {
      if (j > space) {
        process.stdout.write("*");
      } else {
        process.stdout.write(" ");
      }
    }
    space--;
    console.log();
  }
};

// invertedAndRotatedPyramid()

// 1 2 3 4 5
// 1 2 3 4
// 1 2 3
// 1 2
// 1
const decreasingNumberCount = () => {
  let rows = 5;
  for (i = rows; i >= 0; i--) {
    for (j = 1; j <= i; j++) {
      process.stdout.write(`${j} `);
    }
    console.log();
  }
};

// decreasingNumberCount()

// 1
// 2 3
// 4 5 6
// 7 8 9 10
// 11 12 13 14 15

const floydNumber = () => {
  count = 1;
  for (let i = 0; i < 5; i++) {
    for (let j = 0; j <= i; j++) {
      process.stdout.write(`${count} `);
      count++;
    }
    console.log();
  }
};

// floydNumber()

const ZeroOneTriangle = () => {
  for (let i = 1; i <= 5; i++) {
    for (let j = 1; j <= i; j++) {
      if ((i + j) % 2 === 0) {
        process.stdout.write(`1`);
      } else {
        process.stdout.write(`0`);
      }
    }
    console.log();
  }
};

// ZeroOneTriangle();

// col   01234567
//row 0 *      *
//row 1 **    **
//row 2 ***  ***
//row 3 ********
//row 4 ********
//row 5 ***  ***
//row 6 **    **
//row 7 *      *
const butterFly = () => {
  let row = 8;
  let col = 8;
  let half = row / 2;
  let right = 1;
  let left = col - 2;

  for (let i = 0; i < row; i++) {
    for (let j = 0; j < col; j++) {
      if (j < right || j > left) {
        process.stdout.write("*");
      } else {
        process.stdout.write(" ");
      }
    }
    console.log("");

    if (i < half - 1) {
      right++;
      left--;
    } else if (i >= half) {
      right--;
      left++;
    }
  }
};

//   butterFly();

// *****
//  *****
// *****
//  *****
// *****

const rombus = () => {
  size = 5;
  print = 5;
  for (let i = 1; i <= size; i++) {
    for (let j = 1; j <= size * 2 - 1; j++) {
      if (j >= print && j < print + 5) {
        process.stdout.write("*");
      } else {
        process.stdout.write(" ");
      }
    }
    print--;
    console.log();
  }
};

// rombus()

//     *****
//    *   *
//   *   *
//  *   *
// *****
const hollowRombus = () => {
  size = 5;
  print = 5;
  for (let i = 1; i <= size; i++) {
    for (let j = 1; j <= size * 2 - 1; j++) {
      if (j >= print && j < print + 5) {
        if (i === 1 || i == 5 || j == print || j === print + 4) {
          process.stdout.write("*");
        } else {
          process.stdout.write(" ");
        }
      } else {
        process.stdout.write(" ");
      }
    }
    print--;
    console.log();
  }
};

// hollowRombus()

//    *
//   ***
//  *****
// *******
//  *****
//   ***
//    *
const diamond = () => {
  size = 8;
  left = 8 / 2;
  right = 8 / 2;
  for (let i = 1; i < size; i++) {
    for (let j = 1; j < size; j++) {
      if (j >= left && j <= right) {
        process.stdout.write("*");
      } else {
        process.stdout.write(" ");
      }
    }

    if (i >= 4) {
      left++;
      right--;
    } else {
      left--;
      right++;
    }

    console.log();
  }
};

diamond();

const show = () => {
  for (let i = 0; i < 5; i++) {
    for (let j = 0; j < 5; j++) {
      process.stdout.write(` r:${i}, c${j}`);
    }
    console.log();
  }
};

// show()

File: 01-Patterns/python/02-patterns.py
def hollow_box():
    for i in range(5):
        for j in range(5):
            if i == 0 or i == 4 or j == 0 or j == 4:
                print("*", end="")
            else:
                print(" ", end="")
        print()


# hollow_box()


def cross():
    dif = 4
    for i in range(1, 6):
        for j in range(1, 6):
            if i == j or j - i == dif:
                print("*", end="")
            else:
                print(" ", end="")
        print()
        dif -= 2


# cross()


def inverted_and_rotated_pyramid():
    space = 4
    for i in range(5):
        for j in range(5):
            if j > space:
                print("*", end="")
            else:
                print(" ", end="")
        space -= 1
        print()


# inverted_and_rotated_pyramid()



def decreasing_number_count():
    rows = 5
    for i in range(rows, 0, -1):
        for j in range(1, i + 1):
            print(j, end=" ")
        print()


# decreasing_number_count()



def floyd_number():
    count = 1
    for i in range(5):
        for j in range(i + 1):
            print(count, end=" ")
            count += 1
        print()


# floyd_number()


def zero_one_triangle():
    for i in range(1, 6):
        for j in range(1, i + 1):
            if (i + j) % 2 == 0:
                print("1", end="")
            else:
                print("0", end="")
        print()


# zero_one_triangle()


def butterfly():
    row = 8
    col = 8
    half = row // 2
    right = 1
    left = col - 2

    for i in range(row):
        for j in range(col):
            if j < right or j > left:
                print("*", end="")
            else:
                print(" ", end="")
        print("")

        if i < half - 1:
            right += 1
            left -= 1
        elif i >= half:
            right -= 1
            left += 1


# butterfly()


def rhombus():
    size = 5
    print_count = 5
    for i in range(1, size + 1):
        for j in range(1, (size * 2)):
            if j >= print_count and j < print_count + 5:
                print("*", end="")
            else:
                print(" ", end="")
        print_count -= 1
        print()


# rhombus()

def hollow_rhombus():
    size = 5
    print_count = 5
    for i in range(1, size + 1):
        for j in range(1, (size * 2)):
            if j >= print_count and j < print_count + 5:
                if i == 1 or i == 5 or j == print_count or j == print_count + 4:
                    print("*", end="")
                else:
                    print(" ", end="")
            else:
                print(" ", end="")
        print_count -= 1
        print()


# hollow_rhombus()


def diamond():
    size = 8
    left = size // 2
    right = size // 2
    for i in range(1, size):
        for j in range(1, size):
            if j >= left and j <= right:
                print("*", end="")
            else:
                print(" ", end="")

        if i >= 4:
            left += 1
            right -= 1
        else:
            left -= 1
            right += 1

        print()


# diamond()

